frontend/src/utils/transformationExecutor.ts

```diff
@@ -0,0 +1,101 @@
+export interface TransformationResult {
+  success: boolean;
+  document: string | object;
+  error?: string;
+  transformationId?: string;
+}
+
+// Executes a single transformation on a document
+export async function executeTransformation(
+  transformation: string,
+  document: string | object,
+): Promise<TransformationResult> {
+  try {
+    const transformFn = evaluateTransformation(transformation);
+
+    if (typeof transformFn !== "function") {
+      return {
+        success: false,
+        document,
+        error: "Transformation doesn't provide a valid entrypoint",
+      };
+    }
+
+    const result = transformFn(document);
+    return { success: true, document: result };
+  } catch (error) {
+    return {
+      success: false,
+      document,
+      error: error instanceof Error ? error.message : "Unknown error",
+    };
+  }
+}
+
+// A DocumentTransformer is a function that takes a document (object or string)
+// and returns a transformed document (object or string).
+// We need all transformations to conform to this type.
+export type DocumentTransformer = (
+  document: object | string,
+) => object | string;
+
+// Evaluates a transformation code string and returns the entrypoint function
+function evaluateTransformation(code: string): DocumentTransformer | null {
+  try {
+    // Use Function constructor to evaluate the code safely
+    // This can be replaced with isolated-vm or web workers later for better isolation
+    // The way of extracting the entrypoint function is a bit specific to our default transformation format,
+    // this could be broadened.
+    const result = new Function(`
+      ${code}
+      return (document) => {
+        try {
+          const entrypoint = ${code.includes("(() => main)()") ? "(() => main)()" : "null"};
```

Comment:
I might not be reading this correctly, why do we need to look for this exact statement? Shouldn't we try to the execute `code` and then make sure we got back a function that can be used for the transformation?

Reply:
Okay, after discussion with Andre, I loosened this up. The way our transformations are defined doesn't give us an unambiguous definition of "got back a function". So now it looks for this default entrypoint definition and uses that if it can. If not, it takes the last expression of the block and tries to use that as an entrypoint function. That means you can do other approaches like,
```
  function newMain(context) {
    return (document) => {
      document.transformed = true;
      return document;
    };
  }
  newMain;
  ```
  and it will work correctly.